subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java

```diff
@@ -1786,21 +1788,50 @@ private void assertUsageIsMutable() {
         }
     }
 
-    private void logChangingUsage(String usage, boolean allowed) {
-        String msgTemplate = "Allowed usage is changing for %s, %s. Ideally, usage should be fixed upon creation.";
-        if (warnOnChangingUsage) {
-            DeprecationLogger.deprecateBehaviour(String.format(msgTemplate, getDisplayName(), describeChangingUsage(usage, allowed)))
+    private void maybeWarnOnChangingUsage(String usage, boolean current) {
+        if (!isSpecialCaseOfChangingUsage(usage, current)) {
+            String msgTemplate = "Allowed usage is changing for %s, %s. Ideally, usage should be fixed upon creation.";
+            DeprecationLogger.deprecateBehaviour(String.format(msgTemplate, getDisplayName(), describeChangingUsage(usage, current)))
                     .withAdvice("Usage should be fixed upon creation.")
                     .willBeRemovedInGradle9()
                     .withUpgradeGuideSection(8, "configurations_allowed_usage")
                     .nagUser();
-        } else if (LOGGER.isInfoEnabled()) {
-            LOGGER.info(String.format(msgTemplate, getDisplayName(), describeChangingUsage(usage, allowed)));
         }
     }
 
-    private String describeChangingUsage(String usage, boolean allowed) {
-        return usage + " was " + !allowed + " and is now " + allowed;
+    /**
+     * This is a temporary method that decides if a usage change is a known/supported special case, where a deprecation warning message
+     * should not be emitted.
+     * <p>
+     * Many of these exceptions are needed to avoid spamming deprecations warnings whenever the Kotlin plugin is used.  This method is
+     * temporary as the eventual goal is that all configuration usage be specified upon creation and immutable thereafter.
+     * <p>
+     * <ol>
+     *     <li>While {#roleAtCreation} is {@code null}, we are still initializing, so we should NOT warn.</li>
+     *     <li>Changes to the usage of the detached configurations should NOT warn (this done by the Kotlin plugin).</li>
+     *     <li>The legacy role is a special case, and should NOT warn when changing usage (due to the permissiveness of the legacy role, this change will always be a further restriction).</li>
+     *     <li>Setting consumable usage to false on the {@code apiElements} and {@code runtimeElements} configurations should NOT warn (this is done by the Kotlin plugin).</li>
+     *     <li>All other usage changes should warn.</li>
+     * </ol>
+     * <p>
+     * This method is temporary, so the duplication of the configuration names defined in
+     * {@link JavaPlatformPlugin}, which are not available to be referenced directly from here, is unfortunate, but not a showstopper.
```

Comment:
```suggestion
     * {@link JvmConstants}, which are not available to be referenced directly from here, is unfortunate, but not a showstopper.
```

The constants in `JavaPlatformPlugin` are themselves duplicates of those in `JvmConstants`.

Reply:
Yeah, these should be consolidated.  Same problem with `JvmConstants`, but that does seem more appropriate, I'll update it.